// deadcode/lib/traverse.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _traverse = _interopRequireDefault(require("@babel/traverse"));

var _default = function _default(ast) {
  var dependencies = [];
  var dynamicDependencies = [];
  (0, _traverse["default"])(ast, {
    CallExpression: function CallExpression(path) {
      if (isRequireCallee(path.node.callee)) {
        if (path.node.arguments[0].type !== "StringLiteral") {
          dynamicDependencies.push(path.node.arguments[0].loc);
        } else {
          dependencies.push(path.node.arguments[0].value);
        }
      }
    },
    ImportDeclaration: function ImportDeclaration(path) {
      if (path.node.source.type !== "StringLiteral") {
        dynamicDependencies.push(path.node.source.loc);
      } else {
        dependencies.push(path.node.source.value);
      }
    },
    ExportNamedDeclaration: function ExportNamedDeclaration(path) {
      if (path.node.source) {
        if (path.node.source.type !== "StringLiteral") {
          dynamicDependencies.push(path.node.source.loc);
        } else {
          dependencies.push(path.node.source.value);
        }
      }
    }
  });
  return {
    dependencies: dependencies,
    dynamicDependencies: dynamicDependencies
  };
};

exports["default"] = _default;

var isRequireCallee = function isRequireCallee(callee) {
  if (callee.type === "Identifier") {
    return callee.name === "require";
  }

  return false;
};
